p2p/server/msg_server.go

```diff
@@ -0,0 +1,193 @@
+	resHandlers        map[uint64]func(msg []byte)             //response handlers by request ReqId
+	msgRequestHandlers map[MessageType]func(msg []byte) []byte //request handlers by request type
+	ingressChannel     chan service.Message                    //chan to relay messages into the server
+	requestLifetime    time.Duration                           //time a request can stay in the pending queue until evicted
+}
+
+func NewMsgServer(network ServerService, name string, requestLifetime time.Duration) *Message_Server {
+	p := &Message_Server{
+		name:               name,
+		pendingMap:         make(map[uint64]chan interface{}),
+		resHandlers:        make(map[uint64]func(msg []byte)),
+		pendingQueue:       list.New(),
+		network:            network,
+		ingressChannel:     network.RegisterProtocol(name),
+		msgRequestHandlers: make(map[MessageType]func(msg []byte) []byte),
+		requestLifetime:    requestLifetime,
+	}
+
+	go p.readLoop()
+	return p
+}
+
+func (p *Message_Server) readLoop() {
+	for {
+		timer := time.NewTicker(10 * time.Second)
+		select {
+		case <-timer.C:
+			go p.cleanStaleMessages()
+		case msg, ok := <-p.ingressChannel:
+			if !ok {
+				log.Error("read loop channel was closed")
+				break
+			}
+			//todo add buffer and option to limit number of concurrent goroutines
+			go p.handleMessage(msg.(ServerMessage))
+		}
+	}
+}
+
+func (p *Message_Server) cleanStaleMessages() {
+	for {
+		if elem := p.pendingQueue.Front(); elem != nil {
+			item := elem.Value.(Item)
+			if time.Since(item.timestamp) > p.requestLifetime {
+				p.removeFromPending(item.id, elem)
+			} else {
+				return
+			}
+		}
+	}
+}
+
+func (p *Message_Server) handleMessage(msg ServerMessage) {
+	if msg.Data().Req {
+		p.handleRequestMessage(msg.Sender().PublicKey(), msg.Data())
+	} else {
+		p.handleResponseMessage(msg.Data())
+	}
+}
+
+func (p *Message_Server) handleRequestMessage(sender crypto.PublicKey, headers *service.Data_MsgWrapper) {
+
+	if payload := p.msgRequestHandlers[MessageType(headers.MsgType)](headers.Payload); payload != nil {
+		rmsg := &service.Data_MsgWrapper{MsgType: headers.MsgType, ReqID: headers.ReqID, Payload: payload}
+		sendErr := p.network.SendWrappedMessage(sender.String(), p.name, rmsg)
+		if sendErr != nil {
+			log.Error("Error sending response message, err:", sendErr)
+		}
+	}
+}
+
+func (p *Message_Server) handleResponseMessage(headers *service.Data_MsgWrapper) {
+
+	//get and remove from pendingMap
+	p.pendMutex.Lock()
+	pend, okPend := p.pendingMap[headers.ReqID]
+	foo, okFoo := p.resHandlers[headers.ReqID]
+	delete(p.pendingMap, headers.ReqID)
+	delete(p.resHandlers, headers.ReqID)
+	p.pendMutex.Unlock()
+
+	if okPend {
+		if okFoo {
+			foo(headers.Payload)
+		} else {
+			pend <- headers.Payload
+		}
+	}
+}
+
+func (p *Message_Server) removeFromPending(reqID uint64, req *list.Element) {
+	p.pendMutex.Lock()
+	if req != nil {
+		p.pendingQueue.Remove(req)
```

Comment:
there are two objects that are synched here, the pending queue and the maps, these are not always modified together (the queue and maps) so I'd split their mutexes as well